### PR TITLE
Add org.kde.klevernotes

### PR DESCRIPTION
**Pull request by LouisSchul**
### Please confirm your submission meets all the criteria

- [x] Please describe the application briefly.
KleverNotes is a note-taking and management application using markdown.

- [x] The domain used for the application ID is [controlled by the application developers][appid-domain] and the [application id guidelines][appid] are followed.
- [x] I have read the and followed all the [Submission and licence requirements][reqs].
- [x] I have [built][build] and tested the submission locally.
- [x] I am an author/developer/upstream contributor of the project.

[appid-domain]: https://docs.flathub.org/docs/for-app-authors/requirements/#control-over-domain-or-repository
[appid]: https://docs.flathub.org/docs/for-app-authors/requirements#application-id
[reqs]: https://docs.flathub.org/docs/for-app-authors/requirements
[build]: https://docs.flathub.org/docs/for-app-authors/submission/#before-submission

### CODE REVIEW

File: org.kde.klevernotes.json [hunk 1]
```diff
@@ -0,0 +1,61 @@
+{
+    "id": "org.kde.klevernotes",
+    "runtime": "org.kde.Platform",
+    "runtime-version": "6.6",
```
Comment (bbhtt) [inline org.kde.klevernotes.json:4]: 6.7 is the latest, any reason you aren't using it ?

File: org.kde.klevernotes.json [hunk 2]
```diff
@@ -0,0 +1,61 @@
...
+    "sdk": "org.kde.Sdk",
+    "base": "io.qt.qtwebengine.BaseApp",
```
Comment (bbhtt) [inline org.kde.klevernotes.json:6]: you have to cleanup the baseapp, see https://github.com/flathub/io.qt.qtwebengine.BaseApp?tab=readme-ov-file#cleanup

File: org.kde.klevernotes.json [hunk 3]
```diff
@@ -0,0 +1,61 @@
...
+    "base-version": "6.6",
+    "command": "klevernotes",
+    "finish-args": [
+        "--share=ipc",
+        "--socket=x11",
+        "--socket=wayland",
+        "--device=dri",
+        "--filesystem=home",
```
Comment (bbhtt) [inline org.kde.klevernotes.json:14]: why is home needed ?

File: org.kde.klevernotes.json [hunk 4]
```diff
@@ -0,0 +1,61 @@
...
+        "--env=QML_IMPORT_PATH=/app/qml",
+        "--env=QTWEBENGINE_DICTIONARIES_PATH=/app/qtwebengine_dictionaries",
+        "--talk-name=org.freedesktop.Flatpak"
```
Comment (bbhtt) [inline org.kde.klevernotes.json:17]: this needs an exception why is this needed ?

File: org.kde.klevernotes.json (hunk 1 above)
Comment (LouisSchul) [reply org.kde.klevernotes.json:4]: You are right, I did not noticed it, I will bump the version.

File: org.kde.klevernotes.json (hunk 2 above)
Comment (LouisSchul) [reply org.kde.klevernotes.json:6]: Will add this.

File: org.kde.klevernotes.json (hunk 3 above)
Comment (LouisSchul) [reply org.kde.klevernotes.json:14]: I can't switch/create klevernotes storage if I remove this. 
Do you suggest using a different filesystem than `home` ?

File: org.kde.klevernotes.json (hunk 4 above)
Comment (LouisSchul) [reply org.kde.klevernotes.json:17]: It seems like I can remove `"--env=QTWEBENGINE_DICTIONARIES_PATH=/app/qtwebengine_dictionaries"` without any issue.

- `"--env=QML_IMPORT_PATH=/app/qml"` is for the WebEngineView, it won't load without it.
- `"--talk-name=org.freedesktop.Flatpak"` is used to be able to run external program, for example, syntax highlighters.

Comment (bbhtt) [reply org.kde.klevernotes.json:17]: I'm talking about `org.freedesktop.Flatpak`. what syntax highlighters are these and can these be added to the Flatpak package ?

File: org.kde.klevernotes.json (hunk 3 above)
Comment (bbhtt) [reply org.kde.klevernotes.json:14]: what is "klevernotes storage" and where is it created ?

File: flathub.json [hunk 5]
```diff
@@ -0,0 +1,3 @@
+{
+    "only-arches": ["x86_64"]
```
Comment (hfiguiere) [inline flathub.json:2]: why?

**Comment by hfiguiere** [comment]
cc @flathub/kde

File: org.kde.klevernotes.json (hunk 3 above)
Comment (LouisSchul) [reply org.kde.klevernotes.json:14]: It's just a folder. The app need to be able to create and read from a folders/files. The user usually choose the location, that's why I choose `home` snce there's no predefined location.

File: org.kde.klevernotes.json (hunk 4 above)
Comment (LouisSchul) [reply org.kde.klevernotes.json:17]: I'm not sure it could be added as a Flatpak.

It basically rely on external syntax highlighting to do the work, such as [chroma](https://github.com/alecthomas/chroma) or [pygments](https://pygments.org/).

File: flathub.json (hunk 5 above)
Comment (hfiguiere) [reply flathub.json:2]: it sure will never be tested. All `org.kde` is built also on aarch64.

Comment (LouisSchul) [reply flathub.json:2]: Alright, let's remove this.

File: org.kde.klevernotes.json (hunk 4 above)
Comment (LouisSchul) [reply org.kde.klevernotes.json:17]: Those are optionnal, it is up to the user to decide if they want to enable it or not. If they want it, they just install the required package and the option will appear, if not, the app is still completely usable.

Comment (bbhtt) [reply org.kde.klevernotes.json:17]: if it's optional why have this permission by default for everyone ? Like you said those wanting it can add it themselves.

Comment (LouisSchul) [reply org.kde.klevernotes.json:17]: I'm not very familiar with the way flatpak works. 
If this can be added afterward then I'm okay to remove it from this build, no problem.

I think we've covered all the issues with this initial commit, I'll update the PR based on your feedback, thank you.

Comment (hfiguiere) [reply org.kde.klevernotes.json]: > Those are optionnal, it is up to the user to decide if they want to enable it or not. If they want it, they just install the required package and the option will appear, if not, the app is still completely usable.

Enabled vs installed.

My take is that they should be added to the flatpak so that the user can enable. Otherwise there is no point. It's not user friendly.

Comment (LouisSchul) [reply org.kde.klevernotes.json]: I currently allow `chroma`, `pygmentize` and `ksynthaxhighlighter` for the external highlighters to choose from.
I also allow the use of `plantuml` to generate UML graph.

All of those are optionnals and disabled by default. To use them you install the required package and enable the options, that's it.

Having them installed by default would mean that I would need to ship 4 external program on top of KleverNotes inside this flatpak. 
What if I want to support more things in this way in the future ? Even more programs into the flatpak ?

I don't like this idea.

Comment (bbhtt) [reply org.kde.klevernotes.json]: Then don't have this permission by default, users can add it with `flatpak override --talk-name=org.freedesktop.Flatpak` or through Flatseal or I think KDE settings allow overriding permissions as well.

File: org.kde.klevernotes.json [hunk 6]
```diff
@@ -0,0 +1,60 @@
...
+    "runtime": "org.kde.Platform",
+    "runtime-version": "6.7",
+    "sdk": "org.kde.Sdk",
+    "base": "io.qt.qtwebengine.BaseApp",
+    "base-version": "6.6",
+    "command": "klevernotes",
+    "finish-args": [
+        "--share=ipc",
+        "--socket=x11",
+        "--socket=wayland",
```
Comment (bbhtt) [inline org.kde.klevernotes.json:12]: it's either x11 only or wayland+fallback-x11 not both.

**Comment by bbhtt** [comment]
bot, build org.kde.klevernotes

File: org.kde.klevernotes.json [hunk 7]
```diff
@@ -0,0 +1,60 @@
+{
+    "id": "org.kde.klevernotes",
+    "runtime": "org.kde.Platform",
+    "runtime-version": "6.7",
+    "sdk": "org.kde.Sdk",
+    "base": "io.qt.qtwebengine.BaseApp",
+    "base-version": "6.6",
```
Comment (travier) [inline org.kde.klevernotes.json:7]: Shouldn't this be 6.7?

File: org.kde.klevernotes.json (hunk 3 above)
Comment (travier) [reply org.kde.klevernotes.json:14]: The location should be chosen via the file portal where the user would select the directory and then the app would get access to it. I don't think it needs full home access.

File: org.kde.klevernotes.json [hunk 8]
```diff
@@ -0,0 +1,60 @@
...
+    "command": "klevernotes",
+    "finish-args": [
+        "--share=ipc",
+        "--socket=wayland",
+        "--socket=fallback-x11",
+        "--device=dri",
+        "--filesystem=home",
```
Comment (travier) [inline org.kde.klevernotes.json:14]: Let's sort this list.

**Review by travier — CHANGES_REQUESTED** [review]

File: org.kde.klevernotes.json (hunk 7 above)
Comment (bbhtt) [reply org.kde.klevernotes.json:7]: indeed

Comment (bbhtt) [inline org.kde.klevernotes.json:7]:
```suggestion
    "base-version": "6.7",
```

File: org.kde.klevernotes.json [hunk 9]
```diff
@@ -0,0 +1,60 @@
...
+    "runtime": "org.kde.Platform",
+    "runtime-version": "6.7",
+    "sdk": "org.kde.Sdk",
+    "base": "io.qt.qtwebengine.BaseApp",
+    "base-version": "6.7",
+    "command": "klevernotes",
+    "finish-args": [
+        "--device=dri",
+        "--env=QML_IMPORT_PATH=/app/qml",
+        "--filesystem=home",
```
Comment (travier) [inline org.kde.klevernotes.json:12]: Should be removed unless it clearly does not work without (and a bug should be filed then).

Comment (LouisSchul) [reply org.kde.klevernotes.json:12]: You're right, thing seems to work fine without `"--filesystem=home"`. 
Only the displayed path to the storage seems to be affected but that's no big deal.

**Comment by travier** [comment]
bot, build org.kde.klevernotes

**Comment by travier** [comment]
@LouisSchul Note that you have to trigger a build manually here with a comment with `bot, build org.kde.klevernotes` every time you push.

**Comment by bbhtt** [comment]
waiting for someone from flathub/kde to test it, otherwise seems fine.

File: org.kde.klevernotes.json [hunk 10]
```diff
@@ -0,0 +1,59 @@
...
+    "base-version": "6.7",
+    "command": "klevernotes",
+    "finish-args": [
+        "--device=dri",
+        "--env=QML_IMPORT_PATH=/app/qml",
+        "--share=ipc",
+        "--socket=wayland",
+        "--socket=fallback-x11"
+    ],
+    "cleanup-commands": "/app/cleanup-BaseApp.sh",
```
Comment (travier) [inline org.kde.klevernotes.json:16]: We should also clean `include`

**Comment by LouisSchul** [comment]
bot, build org.kde.klevernotes

**Comment by travier** [comment]
So, we've just merged this app, and right away we get errors on the first fedc PR: https://github.com/flathub/org.kde.klevernotes/pull/1

Is there something missing in the validation process here?

**Comment by hfiguiere** [comment]
because f-e-d-c is failing here (or is it anitya?):

image

Note the versions of the main module, that is the one that provide the appstream file fail validation.

**Comment by travier** [comment]
Arg, I missed that. Makes sense. Sorry for the noise.
